src/documents/models.py

```diff
@@ -267,17 +277,90 @@ def __str__(self):
         return str(created)
 
     @property
-    def source_path(self):
-
-        file_name = "{:07}.{}".format(self.pk, self.file_type)
+    def source_filename(self):
+        if self.filename is None:
+            self.filename = self.source_filename_new()
+
+        return self.filename
+
+    def many_to_list(self, field):
+        mylist = []
+        for t in field.all():
+            mylist.append(t.name)
+        return mylist
+
+    def many_to_dictionary(self, field):
+        mydictionary = dict()
+        for t in field.all():
+            delimeter = t.name.find('_')
+
+            if delimeter is -1:
+                continue
+
+            key = t.name[:delimeter]
+            value = t.name[delimeter+1:]
+
+            mydictionary[key] = value
+
+        return mydictionary
+
+    def source_filename_new(self):
+        # Create directory name based on configured format
+        if settings.PAPERLESS_DIRECTORY_FORMAT is not None:
+            directory = settings.PAPERLESS_DIRECTORY_FORMAT.format(
+                        correspondent=self.correspondent,
+                        title=self.title,
+                        created=self.created,
+                        added=self.added,
+                        tags=defaultdict(str,
+                                         self.many_to_dictionary(self.tags)))
+        else:
+            directory = ""
+
+        # Create filename based on configured format
+        if settings.PAPERLESS_FILENAME_FORMAT is not None:
+            filename = settings.PAPERLESS_FILENAME_FORMAT.format(
+                        correspondent=self.correspondent,
+                        title=self.title,
+                        created=self.created,
+                        added=self.added,
+                        tags=defaultdict(str,
+                                         self.many_to_dictionary(self.tags)))
+        else:
+            filename = ""
+
+        path = os.path.join(slugify(directory), slugify(filename))
+
+        # Always append the primary key to guarantee uniqueness of filename
+        if len(path) > 0:
+            filename = "%s-%07i.%s" % (path, self.pk, self.file_type)
+        else:
+            filename = "%07i.%s" % (self.pk, self.file_type)
+
+        # Append .gpg for encrypted files
         if self.storage_type == self.STORAGE_TYPE_GPG:
-            file_name += ".gpg"
+            filename += ".gpg"
+
+        # Create directory for target
+        create_dir = self.filename_to_path(slugify(directory))
+        try:
+            os.makedirs(create_dir)
+        except os.error:
+            # Directory existed already, ignore
+            pass
+
+        return filename
+
+    @property
+    def source_path(self):
+        return self.filename_to_path(self.source_filename)
 
+    def filename_to_path(self, filename):
```

Comment:
Since this doesn't need any class-association, I'd suggest making it a static method. (The call-sites have to be adjusted accordingly.)

```suggestion
    @staticmethod
    def filename_to_path(filename):
```

Reply:
done